File: Part1_강의자료/Ch05_정렬/문제별코드/2910_빈도정렬/2910.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

struct Message {
    int num;
    int idx;
    bool operator<(const Message &o) const{
        if (num != o.num)
            return num < o.num;
        return idx < o.idx;
    }
};

struct Frequency {
    int num;
    int count;
    int first_idx;
    bool operator<(const Frequency &o) const{
        if (count != o.count)
            return count > o.count;
        return first_idx < o.first_idx;
    }
};

int main()
{
    int N, C;
    vector<Message> messages;
    vector<Frequency> frequencies;
    int x;

    scanf("%d %d", &N, &C);
    for (int i = 0; i < N; i++)
    {
        scanf("%d", &x);
        messages.push_back({x, i});
    }

    sort(messages.begin(), messages.end());
    
    frequencies.push_back({messages[0].num, 1, messages[0].idx});
    for (int i = 1; i < N; i++)
    {
        if (messages[i].num != messages[i - 1].num)
            frequencies.push_back({messages[i].num, 0, messages[i].idx});
        frequencies.back().count++;
    }

    sort(frequencies.begin(), frequencies.end());

    for (int i = 0; i < frequencies.size(); i++)
        while (frequencies[i].count--)
            printf("%d ", frequencies[i].num);
    printf("\n");
    return 0;
}

File: Part1_강의자료/Ch05_정렬/문제별코드/2910_빈도정렬/2910.py
class Message:
    def __init__(self, num: int, idx: int):
        self.num = num
        self.idx = idx

class Frequency:
    def __init__(self, num: int, count: int, first_idx: int):
        self.num = num
        self.count = count
        self.first_idx = first_idx
        
N, C = map(int, input().split())
arr = list(map(int, input().split()))
messages = []
frequencies = []

for i in range(N):
    messages.append(Message(arr[i], i))

messages.sort(key=lambda m: (m.num, m.idx))

frequencies.append(Frequency(messages[0].num, 1, messages[0].idx))
for i in range(1, N):
    if messages[i].num != messages[i - 1].num:
        frequencies.append(Frequency(messages[i].num, 0, messages[i].idx))
    frequencies[-1].count += 1

frequencies.sort(key=lambda f: (-f.count, f.first_idx))

for freq in frequencies:
    print((str(freq.num) + ' ') * freq.count, end='')
print()
